### PR TITLE
Course setup

### PR DESCRIPTION
**Pull request by Nimor111**

### CODE REVIEW

File: odin/education/permissions.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+class IsStudentOrTeacherInCoursePermission(BaseUserPassesTestMixin):
+    raise_exception = True
+
+    def test_func(self):
+        course = self.course
+        email = self.request.user.email
+        is_student = course.students.filter(email=email).exists()
+        is_teacher = course.teachers.filter(email=email).exists()
+        is_superuser = self.request.user.is_superuser
```
Comment (RadoRado) [inline odin/education/permissions.py]: This should not be the case for now. We would rather make the superuser a "hidden" teacher to all courses.

**Comment by RadoRado** [comment]
testing slack integration :panda_face:

File: odin/education/permissions.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        self.is_student = course.students.filter(email=email).exists()
+        self.is_teacher = course.teachers.filter(email=email).exists()
+
+        if self.is_student or self.is_teacher:
+            return True and super().test_func()
+
+        return False
+
+
+class IsTeacherInCoursePermission(IsStudentOrTeacherInCoursePermission):
```
Comment (Rositsazz) [inline odin/education/permissions.py]: This permission's logic is the same as in the parent class.

File: odin/education/tests/test_services.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        weeks = course.duration_in_weeks
+        self.assertEqual(1, Course.objects.count())
+        self.assertEqual(weeks, Week.objects.count())
+        week_one = Week.objects.first()
+        self.assertEqual(0, week_one.start_date.weekday())
+
+
+class TestCreateTopic(TestCase):
+
+    def test_create_topic_adds_topic_to_course_successfully(self):
```
Comment (Rositsazz) [inline odin/education/tests/test_services.py]: You can add setUp to these classes

File: odin/templates/dashboard/management.html [hunk 4]
```diff
@@ -14,10 +14,10 @@
                 <div class="caption">
                     <i class="fa fa-comments"></i> User management table </div>
                 <div class="tools">
-                    <a href="javascript:;" class="collapse"> </a>
+                    <a href="javascript:;" class="expand"> </a>
                 </div>
             </div>
-            <div class="portlet-body">
+            <div class="portlet-body" style="display:none;">
```
Comment (Rositsazz) [inline odin/templates/dashboard/management.html]: move this to .css file

File: odin/templates/dashboard/management.html [hunk 5]
```diff
@@ -99,6 +99,44 @@
...
         <!-- END SAMPLE TABLE PORTLET-->
+        <div class="portlet box green">
+            <div class="portlet-title">
+                <div class="caption">
+                    <i class="fa fa-comments"></i> Courses table </div>
+                <div class="tools">
+                    <a href="javascript:;" class="expand"> </a>
+                </div>
+            </div>
+            <div class="portlet-body" style="display:none;">
```
Comment (Rositsazz) [inline odin/templates/dashboard/management.html]: move this to .css file

File: odin/education/models.py [hunk 6]
```diff
@@ -40,6 +41,8 @@ class Course(models.Model):
                                       through='CourseAssignment',
                                       through_fields=('course', 'teacher'))
 
+    slug_url = models.SlugField(null=True, unique=True)
```
Comment (RadoRado) [inline odin/education/models.py]: Every course should have a slug field, so no need for `null=True`

File: odin/education/models.py [hunk 7]
```diff
@@ -136,6 +139,14 @@ class Week(models.Model):
     end_date = models.DateField()
 
     number = models.PositiveIntegerField(default=1)
+    course = models.ForeignKey(Course,
+                               on_delete=models.CASCADE,
+                               related_name='weeks',
+                               blank=True,
```
Comment (RadoRado) [inline odin/education/models.py]: We should not have weeks before course.

File: odin/education/services.py [hunk 8]
```diff
@@ -12,3 +20,75 @@ def add_student(course: Course, student: Student) -> CourseAssignment:
...
+
+
+def create_course(*,
+                  name: str,
+                  start_date: datetime,
+                  end_date: datetime,
+                  repository: str,
+                  facebook_group: str,
+                  video_channel: str,
+                  slug_url: str=None):
```
Comment (RadoRado) [inline odin/education/services.py]: Return type should be `-> Course`

File: odin/education/services.py [hunk 9]
```diff
@@ -12,3 +20,75 @@ def add_student(course: Course, student: Student) -> CourseAssignment:
 
 def add_teacher(course: Course, teacher: Teacher) -> CourseAssignment:
     return CourseAssignment.objects.create(course=course, teacher=teacher)
+
+
+def create_course(*,
```
Comment (RadoRado) [inline odin/education/services.py]: If we are doing more than 1 DB write in a service, it's good idea to add `@transaction.atomic` decorator.

This is from `from django.db import transaction`

File: odin/education/services.py [hunk 10]
```diff
@@ -12,3 +20,75 @@ def add_student(course: Course, student: Student) -> CourseAssignment:
...
+        video_channel=video_channel,
+        slug_url=slug_url
+    )
+
+    weeks = course.duration_in_weeks
+    start_date = course.start_date
+    start_date = start_date - timedelta(days=start_date.weekday())
+
+    for i in range(1, weeks + 1):
+        Week.objects.create(course=course,
```
Comment (RadoRado) [inline odin/education/services.py]: I think we can do a batch create here?

File: odin/education/services.py [hunk 11]
```diff
@@ -12,3 +20,75 @@ def add_student(course: Course, student: Student) -> CourseAssignment:
...
+                            end_date=start_date + timedelta(days=7))
+        start_date = start_date + timedelta(days=7)
+
+    return course
+
+
+def create_topic(*,
+                 name: str,
+                 week: Week,
+                 course: Course):
```
Comment (RadoRado) [inline odin/education/services.py]: Return type should be `-> Topic`

File: odin/education/services.py [hunk 12]
```diff
@@ -12,3 +20,75 @@ def add_student(course: Course, student: Student) -> CourseAssignment:
...
+
+    return topic
+
+
+def create_included_material(*,
+                             identifier: str=None,
+                             url: str=None,
+                             topic: Topic=None,
+                             content: str=None,
+                             material: Material=None):
```
Comment (RadoRado) [inline odin/education/services.py]: Return type should be `-> IncludedMaterial`

File: odin/education/urls.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
+from django.conf.urls import url, include
+
+
+from .views import (UserCoursesView,
```
Comment (RadoRado) [inline odin/education/urls.py]: Imports like that should start on new line after.

File: odin/education/views.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+                     week=form.cleaned_data.get('week'),
+                     course=self.course)
+
+        return super().form_valid(form)
+
+
+class AddIncludedMaterialFromExistingView(CourseViewMixin,
+                                          LoginRequiredMixin,
+                                          IsTeacherInCoursePermission,
+                                          ListView):
```
Comment (RadoRado) [inline odin/education/views.py]: Why is this a `ListView`?

File: odin/education/views.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        topic = get_object_or_404(Topic, id=self.kwargs.get('topic_id'))
+        material = get_object_or_404(Material, identifier=request.POST.get('material_identifier'))
+
+        data = {}
+        data['topic'] = topic.id
+        data['material'] = material.id
+
+        form = IncludedMaterialFromExistingForm(data)
+
+        if form.is_valid():
```
Comment (RadoRado) [inline odin/education/views.py]: What happens if the form is invalid?

File: odin/education/views.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        return reverse_lazy('dashboard:education:user-course-detail',
+                            kwargs={'course_id': self.course.id})
+
+    def get_initial(self):
+        self.initial = super().get_initial()
+        self.initial['topic'] = self.kwargs.get('topic_id')
+        return self.initial.copy()
+
+    def get_form_kwargs(self):
+        form_kwgs = super().get_form_kwargs()
```
Comment (RadoRado) [inline odin/education/views.py]: `kwgs` -> `kwargs`

File: odin/management/tests/test_views.py [hunk 17]
```diff
@@ -205,13 +219,18 @@ def test_post_is_forbidden_when_user_is_not_superuser(self):
             self.assertEqual(403, response.status_code)
 
     def test_post_creates_teacher_when_user_is_superuser(self):
-        with self.login(email=self.user.email, password=self.test_password):
-            self.assertEqual(0, Teacher.objects.count())
+        user = BaseUserFactory(password=self.test_password)
```
Comment (RadoRado) [inline odin/management/tests/test_views.py]: There is a `SuperUserFactory` for this.

File: odin/education/factories.py [hunk 18]
```diff
@@ -18,12 +19,56 @@ class Meta:
...
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        return create_course(*args, **kwargs)
+
+
+class WeekFactory(factory.DjangoModelFactory):
+    number = factory.LazyAttribute(lambda _: faker.pyint())
+
+    start_date = factory.LazyAttribute(lambda _: faker.date())
```
Comment (RadoRado) [inline odin/education/factories.py]: This should be `.date_object()` since `.date()` returns a string

Comment (RadoRado) [inline odin/education/factories.py]: Same applies for all factories that need to produce a date.

**Pull request merged**